Add latest changes for kruize object and API's

bharathappali opened the pull request.

dinogun commented on src/main/java/com/autotune/utils/AnalyzerConstants.java:
```diff
@@ -420,4 +420,11 @@ public static final class PerformanceProfileConstants {
+        DEPLOYMENT,
+        STATEFULSET,
+        REPLICASET,
+        POD
```
The initial supported list should be
```
Deployment
StatefulSet
ReplicaSet
ReplicationController
DaemonSet
```

bharathappali replied:
Sure will change them.

bharathappali commented on src/main/java/com/autotune/utils/AnalyzerConstants.java:
```diff
@@ -420,4 +420,31 @@ public static final class PerformanceProfileConstants {
+
+            }
+
+            public static final String DEPLOYMENT = "deployment";
```
@dinogun Can I please know if these are the right formats to compare the k8s object types which we receive in JSON? Please lemme know if I need to make any casing changes.

dinogun replied:
@kusumachalasani can comment what is returned by the queries and we can keep it to be the same as those returned by the queries to avoid having to manipulate it later

kusumachalasani replied:
@bharathappali Below are the workloads and their format(based on the output from queries) that are supported for MVP.
```
deployment
deploymentconfig
daemonset
statefulset
replicaset
replicationcontroller
```

bharathappali replied:
Thanks @kusumachalasani 

@dinogun I have implemented `equals ignore case` to confirm the match. Should we implement a strict match?
I'm more inclined towards `equals ignore case` coz things may change in future and the code still works unaffected.

bharathappali commented:
@dinogun Here are the updates made to the API's `Create Experiment`, `Update Results`, `List Recommendations`. Please lemme know if any other changes needed.

##### Create Experiment:

*API Path*
`http://192.168.49.2:30960/createExperiment`

*Request:*
```
[
    {
  "version": "1.0",
  "experiment_name": "quarkus-resteasy-autotune-min-http-response-time-db",
  "cluster_name": "cluster-one-division-bell",
  "performance_profile": "resource-optimization-openshift",
  "mode": "monitor",
  "target_cluster": "remote",
  "kubernetes_objects": [
    {
      "type": "deployment",
      "name": "tfb-qrh-deployment",
      "namespace": "default",
      "containers": [
        {
          "container_image_name": "kruize/tfb-db:1.15",
          "container_name": "tfb-server-0"
        },
        {
          "container_image_name": "kruize/tfb-qrh:1.13.2.F_et17",
          "container_name": "tfb-server-1"
        }
      ]
    }
  ],
  "trial_settings": {
    "measurement_duration": "15min"
  },
  "recommendation_settings": {
    "threshold": "0.1"
  }
}
]
```

*Response:*
```
{
    "message": "Experiment registered successfully with Kruize. View registered experiments at /listExperiments",
    "httpcode": 201,
    "documentationLink": "",
    "status": "SUCCESS"
}
```

##### Update Results:

*API Path*
`http://192.168.49.2:30960/updateResults`

*Request:*
```
[
  {
    "version": "1.0",
    "experiment_name": "quarkus-resteasy-autotune-min-http-response-time-db",
    "start_timestamp": "2022-01-23T18:25:43.511Z",
    "end_timestamp": "2022-01-23T18:25:43.511Z",
    "kubernetes_objects": [
      {
        "type": "deployment",
        "name": "tfb-qrh-deployment",
        "namespace": "default",
        "containers": [
          {
            "container_image_name": "kruize/tfb-qrh:1.13.2.F_et17",
            "container_name": "tfb-server-0",
            "metrics": [
              {
                "name": "cpuRequest",
                "results": {
                  "value": 1.1,
                  "format": "cores",
                  "aggregation_info": {
                    "sum": 4.4,
                    "avg": 1.1,
                    "format": "cores"
                  }
                }
              },
              {
                "name": "cpuLimit",
                "results": {
                  "value": 0.5,
                  "format": "cores",
                  "aggregation_info": {
                    "sum": 2.0,
                    "avg": 0.5,
                    "format": "cores"
                  }
                }
              },
              {
                "name": "cpuUsage",
                "results": {
                  "value": 0.12,
                  "format": "cores",
                  "aggregation_info": {
                    "min": 0.14,
                    "max": 0.84,
                    "sum": 0.84,
                    "avg": 0.12,
                    "format": "cores"
                  }
                }
              },
              {
                "name": "cpuThrottle",
                "results": {
                  "value": 0.045,
                  "format": "cores",
                  "aggregation_info": {
                    "sum": 0.19,
                    "max": 0.09,
                    "avg": 0.045,
                    "format": "cores"
                  }
                }
              },
              {
                "name": "memoryRequest",
                "results": {
                  "value": 50.12,
                  "format": "MiB",
                  "aggregation_info": {
                    "sum": 250.85,
                    "avg": 50.21,
                    "format": "MiB"
                  }
                }
              },
              {
                "name": "memoryLimit",
                "results": {
                  "value": 100,
                  "format": "MiB",
                  "aggregation_info": {
                    "sum": 500,
                    "avg": 100,
                    "format": "MiB"
                  }
                }
              },
              {
                "name": "memoryUsage",
                "results": {
                  "value": 40.1,
                  "format": "MiB",
                  "aggregation_info": {
                    "min": 50.6,
                    "max": 198.50,
                    "sum": 198.50,
                    "avg": 40.1,
                    "format": "MiB"
                  }
                }
              },
              {
                "name": "memoryRSS",
                "results": {
                  "aggregation_info": {
                    "min": 50.6,
                    "max": 123.6,
                    "sum": 123.6,
                    "avg": 31.91,
                    "format": "MiB"
                  }
                }
              }
            ]
          },
          {
            "container_image_name": "kruize/tfb-qrh:1.13.2.F_et17",
            "container_name": "tfb-server-1",
            "metrics": [
              {
                "name": "cpuRequest",
                "results": {
                  "value": 1.1,
                  "format": "cores",
                  "aggregation_info": {
                    "sum": 4.4,
                    "avg": 1.1,
                    "format": "cores"
                  }
                }
              },
              {
                "name": "cpuLimit",
                "results": {
                  "value": 0.5,
                  "format": "cores",
                  "aggregation_info": {
                    "sum": 2.0,
                    "avg": 0.5,
                    "format": "cores"
                  }
                }
              },
              {
                "name": "cpuUsage",
                "results": {
                  "value": 0.12,
                  "format": "cores",
                  "aggregation_info": {
                    "min": 0.14,
                    "max": 0.84,
                    "sum": 0.84,
                    "avg": 0.12,
                    "format": "cores"
                  }
                }
              },
              {
                "name": "cpuThrottle",
                "results": {
                  "value": 0.045,
                  "format": "cores",
                  "aggregation_info": {
                    "sum": 0.19,
                    "max": 0.09,
                    "avg": 0.045,
                    "format": "cores"
                  }
                }
              },
              {
                "name": "memoryRequest",
                "results": {
                  "value": 50.12,
                  "format": "MiB",
                  "aggregation_info": {
                    "sum": 250.85,
                    "avg": 50.21,
                    "format": "MiB"
                  }
                }
              },
              {
                "name": "memoryLimit",
                "results": {
                  "value": 100,
                  "format": "MiB",
                  "aggregation_info": {
                    "sum": 500,
                    "avg": 100,
                    "format": "MiB"
                  }
                }
              },
              {
                "name": "memoryUsage",
                "results": {
                  "value": 40.1,
                  "format": "MiB",
                  "aggregation_info": {
                    "min": 50.6,
                    "max": 198.50,
                    "sum": 198.50,
                    "avg": 40.1,
                    "format": "MiB"
                  }
                }
              },
              {
                "name": "memoryRSS",
                "results": {
                  "aggregation_info": {
                    "min": 50.6,
                    "max": 123.6,
                    "sum": 123.6,
                    "avg": 31.91,
                    "format": "MiB"
                  }
                }
              }
            ]
          }
        ]
      }
    ]
  }
]
```

*Response:*
```
{
    "message": "Results added successfully! View saved results at /listExperiments.",
    "httpcode": 201,
    "documentationLink": "",
    "status": "SUCCESS"
}
```

##### List Recommendation

*API Path:*
`http://192.168.49.2:30960/listRecommendations`

*Response:*
```
[
    {
        "kubernetes_objects": [
            {
                "type": "deployment",
                "name": "tfb-qrh-deployment",
                "namespace": "default",
                "containers": [
                    {
                        "container_image_name": "kruize/tfb-qrh:1.13.2.F_et17",
                        "container_name": "tfb-server-1",
                        "recommendation": {
                            "2022-01-23T18:25:43.511Z": {
                                "Short Term": {
                                    "monitoringStartTime": "2022-01-22T18:25:43.511Z",
                                    "monitoringEndTime": "2022-01-23T18:25:43.511Z",
                                    "duration_in_hours": 0.0,
                                    "podsCount": 0,
                                    "confidence_level": 0.0,
                                    "config": {
                                        "max": {
                                            "memory": {
                                                "amount": 982.5997506234414,
                                                "format": ""
                                            },
                                            "cpu": {
                                                "amount": 6.51,
                                                "format": ""
                                            }
                                        },
                                        "capacity": {
                                            "memory": {
                                                "amount": 123.6,
                                                "format": ""
                                            },
                                            "cpu": {
                                                "amount": 1.03,
                                                "format": ""
                                            }
                                        }
                                    }
                                },
                                "Middle Term": {
                                    "podsCount": 0,
                                    "confidence_level": 0.0,
                                    "errorMessage": "There is not enough data available to generate a recommendation."
                                },
                                "Long Term": {
                                    "podsCount": 0,
                                    "confidence_level": 0.0,
                                    "errorMessage": "There is not enough data available to generate a recommendation."
                                }
                            }
                        }
                    },
                    {
                        "container_image_name": "kruize/tfb-db:1.15",
                        "container_name": "tfb-server-0",
                        "recommendation": {
                            "2022-01-23T18:25:43.511Z": {
                                "Short Term": {
                                    "monitoringStartTime": "2022-01-22T18:25:43.511Z",
                                    "monitoringEndTime": "2022-01-23T18:25:43.511Z",
                                    "duration_in_hours": 0.0,
                                    "podsCount": 0,
                                    "confidence_level": 0.0,
                                    "config": {
                                        "max": {
                                            "memory": {
                                                "amount": 982.5997506234414,
                                                "format": ""
                                            },
                                            "cpu": {
                                                "amount": 6.51,
                                                "format": ""
                                            }
                                        },
                                        "capacity": {
                                            "memory": {
                                                "amount": 123.6,
                                                "format": ""
                                            },
                                            "cpu": {
                                                "amount": 1.03,
                                                "format": ""
                                            }
                                        }
                                    }
                                },
                                "Middle Term": {
                                    "podsCount": 0,
                                    "confidence_level": 0.0,
                                    "errorMessage": "There is not enough data available to generate a recommendation."
                                },
                                "Long Term": {
                                    "podsCount": 0,
                                    "confidence_level": 0.0,
                                    "errorMessage": "There is not enough data available to generate a recommendation."
                                }
                            }
                        }
                    }
                ]
            }
        ],
        "version": "1.0",
        "experiment_name": "quarkus-resteasy-autotune-min-http-response-time-db"
    }
]
```

dinogun commented:
@bharathappali looks like the API.md changes are still pending.

bharathappali commented:
@dinogun Oh yes I need to add them, I'm making changes for the `snake_case` will be adding doc changes along with that.

bharathappali commented:
@dinogun Thanks for the reviews, I have made changes to convert the JSON keys to `snake_case` and have added the documentation for the API's. Please lemme know if any other changes needed. 

Thanks in advance!

bharathappali commented:
Kruize demo log with these changes
[demo_log.txt](https://github.com/kruize/autotune/files/10848394/demo_log.txt)

dinogun commented:
@bharathappali The API.md changes need to be done in [MonitoringModeAPI.md](https://github.com/kruize/autotune/blob/6dbe449e40f1982dc41625f59475400bae02d7a1/design/MonitoringModeAPI.md)

bharathappali commented:
@dinogun Thanks for pointing it out, I was in a different impression earlier about the doc. I have reverted my earlier doc changes and added the changes to MonitoringModeAPI.md. The PR can be reviewed now.

dinogun reviewed (approved):
LGTM

dinogun commented:
The build error seems to be a rate limiting issue on dockerhub. Will merge as the previous build succeeded and the subsequent changes are only doc related ones
